### PR TITLE
Perps

### PR DESCRIPTION
**Pull request by filipzeta**

### CODE REVIEW

File: examples/cranking/cranking-example.ts [hunk 1]
```diff
@@ -29,7 +29,7 @@ switch (process.env["network"]) {
 
 const NETWORK_URL = process.env["network_url"]!;
 const PROGRAM_ID = new PublicKey(process.env["program_id"]);
-let crankingMarkets = new Array(constants.ACTIVE_MARKETS).fill(false);
+let crankingMarkets = new Array(constants.ACTIVE_MARKETS - 1).fill(false);
```
Comment (xjcaa) [inline examples/cranking/cranking-example.ts:32]: is this right? active markets is 46 right, total markets is 138?

Comment (filipzeta) [reply examples/cranking/cranking-example.ts:32]: I added + 1 to ACTIVE_MARKETS for perps

File: src/utils.ts [hunk 2]
```diff
@@ -708,6 +723,7 @@ export async function processTransaction(
     );
     return txSig;
   } catch (err) {
+    console.log(err);
```
Comment (filipzeta) [inline src/utils.ts:726]: - [ ] remove before merging

File: examples/cranking/cranking-example.ts [hunk 3]
```diff
@@ -29,7 +29,7 @@ switch (process.env["network"]) {
 
 const NETWORK_URL = process.env["network_url"]!;
```
Comment (xjcaa) [inline examples/cranking/cranking-example.ts:30]: This also needs to crank the perp market?

File: src/client.ts [hunk 4]
```diff
@@ -357,14 +358,14 @@ export class Client {
...
-        Exchange.getSubExchange(asset).markets.markets[market].address;
+      marketPubkey = Exchange.getMarket(asset, market).address;
     } else {
       marketPubkey = market;
     }
     return marketPubkey;
   }
 
+  // TODO giving Exchange.getPerpMarket(asset).marketIndex doesn't work here, fix
   public async placeOrder(
```
Comment (xjcaa) [inline src/client.ts:369]: I think just use a separate api man - > placePerpOrder(asset, ...)

File: src/constants.ts [hunk 5]
```diff
@@ -26,7 +27,7 @@ export const MAX_SET_REFERRALS_REWARDS_ACCOUNTS = 12;
 export const MARKET_INDEX_LIMIT = 18;
 // 3 accounts per set * 9 = 27 + 2 = 29 accounts.
 export const CLEAN_MARKET_LIMIT = 9;
-export const CRANK_ACCOUNT_LIMIT = 12;
+export const CRANK_ACCOUNT_LIMIT = 10;
```
Comment (xjcaa) [inline src/constants.ts:30]: can we have diff limit for perp and normal

File: src/risk.ts [hunk 6]
```diff
@@ -19,18 +19,27 @@ import {
```
Comment (xjcaa) [inline src/risk.ts:19]: This file should handle the net funding pnl of the position vs greeks as wel

**Review by kelvinlau20100 — APPROVED** [review]
lfg

**Pull request merged**